pytest_socket.py

```diff
@@ -62,3 +78,47 @@ def enable_socket():
     """ re-enable socket.socket to enable the Internet. useful in testing.
     """
     socket.socket = _true_socket
+
+
+def pytest_configure(config):
+    config.addinivalue_line("markers", "disable_socket(): Disable socket connections for a specific test")
+    config.addinivalue_line("markers", "enable_socket(): Enable socket connections for a specific test")
+    config.addinivalue_line("markers", "restrict_hosts([hosts]): Restrict socket connection to defined list of hosts")
+
+
+def pytest_runtest_setup(item):
+    mark_restrictions = item.get_closest_marker('restrict_hosts')
+    cli_restrictions = item.config.getoption('--restrict-hosts')
+    hosts = None
+    if mark_restrictions:
+        hosts = mark_restrictions.args[0]
+    elif cli_restrictions:
+        hosts = cli_restrictions
+    socket_restrict_hosts(hosts)
+
+
+def pytest_runtest_teardown():
+    remove_host_restrictions()
+
+
+def socket_restrict_hosts(allowed=None):
+    """ disable socket.socket.connect() to disable the Internet. useful in testing.
+    """
+    if not allowed:
+        return
+    if isinstance(allowed, str):
+        allowed = allowed.split(',')
+
+    def guarded_connect(inst, address):
+        host, port = address
```

Comment:
If you don't use `port`, is it useful to parse/pass it around, or should we go with `host` only for the first iteration?

Reply:
it's a tuple, so it's just a way of getting the host. i could have just done `host = address[0]`, which i can do if it's an issue.